### PR TITLE
Observable source asset decorator

### PR DESCRIPTION
**Pull request by smackesey**
### Summary & Motivation

New `observable_source_asset` decorator.

### How I Tested These Changes

Unit tests

### CODE REVIEW

**Comment by smackesey** [comment]
Current dependencies on/for this PR:
* master
  * **PR #9899** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/9899" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/9899?utm_source=stack-comment).

File: python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    description: Optional[str] = None,
+    partitions_def: Optional[PartitionsDefinition] = None,
+    _metadata_entries: Optional[Sequence[Union[MetadataEntry, PartitionMetadataEntry]]] = None,
+    group_name: Optional[str] = None,
+    resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
+) -> "_SourceAsset":
+    ...
+
+
+def source_asset(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py:52]: what do you think about calling this something like `observeable_source_asset`?   I've thought about the idea of a different source asset decorator that contains a function that loads the source asset (which could then maybe be called `loadable_source_asset`).

File: python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            key=source_asset_key,
+            metadata=self.metadata,
+            io_manager_key=self.io_manager_key,
+            io_manager_def=self.io_manager_def,
+            description=self.description,
+            partitions_def=self.partitions_def,
+            _metadata_entries=self._metadata_entries,
+            group_name=self.group_name,
+            resource_defs=self.resource_defs,
+            node_def=op,
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py:143]: I suspect that there are situations where it may be useful inside a sensor instead of inside a run.  E.g. perhaps the sensor produced by `build_asset_reconciliation_sensor` should invoke it as part of sensor execution.

1. Thoughts?
2. If so, then maybe the SourceAsset shouldn't have an op and should just have this function directly? Then we can convert it to an op later on.

File: python_modules/dagster/dagster/_core/definitions/logical_version.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+from typing_extensions import Final
+
+from dagster import _check as check
+
+if TYPE_CHECKING:
+    from dagster._core.definitions.events import AssetKey
+    from dagster._core.events.log import EventLogEntry
+    from dagster._core.instance import DagsterInstance
+
+class LogicalVersion:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/logical_version.py:15]: Do you imagine we may later want to add other attributes to this?

File: python_modules/dagster/dagster/_core/definitions/logical_version.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        else get_most_recent_logical_version(k, key_to_is_source_map[k], instance)
+        for k in ordered_dependency_keys
+    ]
+    all_inputs = [op_version, *dependency_logical_versions]
+    hash_sig = sha256()
+    hash_sig.update(bytearray("".join(all_inputs), "utf8"))
+    return LogicalVersion(hash_sig.hexdigest())
+
+
+def get_most_recent_logical_version(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/logical_version.py:55]: I would expect this to live inside the instance.

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/logical_version.py:55]: also, can we do some caching or something to avoid fetching the same event multiple times within a single step?

File: python_modules/dagster/dagster/_core/definitions/source_asset.py [hunk 5]
```diff
@@ -126,6 +161,46 @@ def io_manager_def(self) -> Optional[IOManagerDefinition]:
...
+    def is_versioned(self) -> bool:
+        return self.node_def is not None
+
+    @property
+    def node_def(self):
+        """Op that generates observation metadata for a source asset."""
+        if self.observe_fn is None:
+            return None
+        else:
+            observe_fn = self.observe_fn
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/source_asset.py:184]: Is this line necessary? Why not just use `self.observe_fn` below?

File: python_modules/dagster/dagster/_core/definitions/source_asset.py [hunk 6]
```diff
@@ -126,6 +161,46 @@ def io_manager_def(self) -> Optional[IOManagerDefinition]:
...
+        )
+        return cast(OpDefinition, self.node_def)
+
+    @public
+    @property
+    def is_versioned(self) -> bool:
+        return self.node_def is not None
+
+    @property
+    def node_def(self):
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/source_asset.py:179]: Return type annotation would be helpful

File: python_modules/dagster/dagster/_core/definitions/source_asset.py [hunk 7]
```diff
@@ -200,3 +277,15 @@ def get_resource_requirements(self) -> Iterator[ResourceRequirement]:
         )
         for source_key, resource_def in self.resource_defs.items():
             yield from resource_def.get_resource_requirements(outer_context=source_key)
+
+    def __hash__(self):
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/source_asset.py:281]: This strikes me as a tiny bit sketchy. Do we anticipate using source assets as keys in dictionaries?

File: python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    key_prefix: Optional[CoercibleToAssetKeyPrefix] = None,
+    metadata: Optional[MetadataUserInput] = None,
+    io_manager_key: Optional[str] = None,
+    io_manager_def: Optional[IOManagerDefinition] = None,
+    description: Optional[str] = None,
+    partitions_def: Optional[PartitionsDefinition] = None,
+    _metadata_entries: Optional[Sequence[Union[MetadataEntry, PartitionMetadataEntry]]] = None,
+    group_name: Optional[str] = None,
+    resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
+) -> Union[SourceAsset, "_ObservableSourceAsset"]:
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py:55]: This needs docstring

File: python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    description: Optional[str] = None,
+    partitions_def: Optional[PartitionsDefinition] = None,
+    _metadata_entries: Optional[Sequence[Union[MetadataEntry, PartitionMetadataEntry]]] = None,
+    group_name: Optional[str] = None,
+    resource_defs: Optional[Mapping[str, ResourceDefinition]] = None,
+) -> "_ObservableSourceAsset":
+    ...
+
+
+def observable_source_asset(
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py:42]: I think it would be safest to mark this as experimental for now, though open to otherwise if you disagree

File: python_modules/dagster/dagster/_core/definitions/source_asset.py [hunk 10]
```diff
@@ -24,27 +27,44 @@
...
-            ("description", PublicAttr[Optional[str]]),
-            ("partitions_def", PublicAttr[Optional[PartitionsDefinition]]),
-            ("group_name", PublicAttr[str]),
-            ("resource_defs", PublicAttr[Dict[str, ResourceDefinition]]),
-        ],
-    ),
-    ResourceAddable,
-):
+
+class SourceAssetObserveFunctionWithContext(Protocol):
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/source_asset.py:44]: Why not use `Callable[[SourceAssetObserveContext], MetadataUserInput`?

File: python_modules/dagster/dagster/_core/definitions/source_asset.py [hunk 11]
```diff
@@ -58,10 +78,22 @@ class SourceAsset(
...
+    key: PublicAttr[AssetKey]
+    metadata_entries: Sequence[Union[MetadataEntry, PartitionMetadataEntry]]
+    io_manager_key: PublicAttr[Optional[str]]
+    _io_manager_def: PublicAttr[Optional[IOManagerDefinition]]
+    description: PublicAttr[Optional[str]]
+    partitions_def: PublicAttr[Optional[PartitionsDefinition]]
+    group_name: PublicAttr[str]
+    resource_defs: PublicAttr[Dict[str, ResourceDefinition]]
+    observe_fn: PublicAttr[Optional[SourceAssetObserveFunction]]
+    _node_def: Optional[OpDefinition]
```
Comment (sryza) [inline python_modules/dagster/dagster/_core/definitions/source_asset.py:93]: I think worth adding a comment saying that this is computed lazily

**Review by sryza — COMMENTED** [review]
Left a few comments but looks pretty much there

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 5 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:186]: This is required to create a closure for the observe function-- mypy flags it otherwise. While I don't _think_ it would break using `self.observe_fn` directly, the type-checker correctly observes that `self.observe_fn`, being a nullable property, might return `None` when it's called at some unpredictable point inside the `compute_fn`.

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 7 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:283]: This was added to allow de-duping of source assets inside sets in a downstream PR. Might no longer be required, I'll remove it from this and add it back if it turns out to be needed downsteram.

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 10 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:46]: Because the function only optionally takes a context function. This can't be represented with `Callable` unfortunately (that's why we don't do it for regular assets).

File: python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py (hunk 9 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/decorators/source_asset_decorator.py:42]: agreed

File: python_modules/dagster-graphql/dagster_graphql/cli.py [hunk 12]
```diff
@@ -43,12 +43,15 @@ def execute_query(
 
     context = workspace_process_context.create_request_context()
 
-    result = cast(ExecutionResult, graphql(
-        request_string=query,
-        schema=create_schema(),
-        context_value=context,
-        variable_values=variables,
-    ))
+    result = cast(
```
Comment (sryza) [inline python_modules/dagster-graphql/dagster_graphql/cli.py:46]: This looks unrelated, right?

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 10 above)
Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:48]: I was imagining:
```
SourceAssetObserveFunction: TypeAlias = Union[
    Callable[[SourceAssetObserveContext], MetadataUserInput], Callable[[], MetadataUserInput]
]

```

**Review by sryza — APPROVED** [review]
A couple small last things. Otherwise looks great.

File: python_modules/dagster/dagster/_core/definitions/source_asset.py (hunk 10 above)
Comment (smackesey) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:48]: Yeah, this was the first thing I tried and for some reason it does not work (though logically it should).

Comment (sryza) [reply python_modules/dagster/dagster/_core/definitions/source_asset.py:48]: got it - was not a blocking concern, just curious
